Add script for source integrity checking

dgreen-arm opened the pull request:
## Description
This script checks for undesired changes to the source code, including permission changes, presence of tabs, non-Unix line endings, trailing whitespace, and TODO comments.

## Status
**READY**

## Requires Backporting
NO

sbutcher-arm commented on scripts/integrity_check.py:
```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/env python3
+
```
Please add the usual copyright and license statement at the top.

gilles-peskine-arm commented on scripts/integrity_check.py:
```diff
@@ -0,0 +1,141 @@
+        self.setup_logger(log_file)
+        self.source_files_to_check = (
+            ".c", ".h", ".function", ".data",
+            ".md", "Makefile", "CMakeLists.txt"
```
Please check `ChangeLog` as well.

dgreen-arm replied:
Added.

gilles-peskine-arm commented on scripts/integrity_check.py:
```diff
@@ -0,0 +1,141 @@
+                    self.trailing_whitespace["files"][filepath].append(i + 1)
+                if not allow_tabs and "\t" in line:
+                    self.tab_issues["files"][filepath].append(i + 1)
+                if "TODO" in line:
```
How does this script cope with itself? It contains `TODO`.

dgreen-arm replied:
The script originally wasn't checking `.py` files, so it didn't check itself. I've added an exemption in the rewrite.

gilles-peskine-arm commented on scripts/integrity_check.py:
```diff
@@ -0,0 +1,141 @@
+#
+# This script requires python 3. This script checks for undesired changes
+# to the source code, including file permission changes, presence of tabs,
+# non-Unix line endings, trailing whitespace, and TODO comments.
```
I have new a feature request: check for non-ASCII characters. There are currently no non-ASCII characters in our code base.

If we decide to allow non-ASCII characters, then please at the very least ensure that files don't start with `\xef\xbb\xbf` (the UTF-8 “BOM”). Yes, I'm asking for it because it happened and it isn't visually apparent.

gilles-peskine-arm replied:
Darryl let me know that I'd flubbed my check, we do have comments with people's names containing non-ASCII characters. So I'm revising my request to just forbidding the UTF-8 BOM.

hanno-arm replied:
@gilles-peskine-arm @dgreen-arm If I run the script on a file saved as `UTF-16` in `gedit`, the script chokes and fails with 
```
hanno@wuehlmaus:~/mbedtlsX$ ./scripts/integrity_check.py 
Traceback (most recent call last):
  File "./scripts/integrity_check.py", line 217, in <module>
    run_main()
  File "./scripts/integrity_check.py", line 212, in run_main
    integrity_check.check_files()
  File "./scripts/integrity_check.py", line 190, in check_files
    issue_to_check.check_file_for_issue(filepath)
  File "./scripts/integrity_check.py", line 84, in check_file_for_issue
    if not f.read().endswith("\n"):
  File "/usr/lib/python3.5/codecs.py", line 321, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 0: invalid start byte
```

gilles-peskine-arm replied:
You need to open the file in binary. Python 3 supposedly makes this easier than Python 2.

dgreen-arm replied:
I can change the script to always open the file in binary to avoid these issues. Would we want to check for the UTF-16 and UTF-32 BOM's as well?

gilles-peskine-arm replied:
If a file was in UTF-16 or UTF-32, then a BOM would be desirable. I suppose we could check that no file that's supposed to be a text file contains a null byte, or more generally control characters not among `\n\r\t\f`. That would in practice detect any file saved as UTF-16 or UTF-32. I don't think we've ever had this particular problem, but we might as well check this while we're at it.

dgreen-arm commented:
I've done a rewrite of the script to make it easier to do any future changes or improvements.

gilles-peskine-arm commented on scripts/integrity_check.py:
```diff
@@ -0,0 +1,225 @@
+#!/usr/bin/env python3
```
I don't like the name ”integrity check“. This is a cryptography library and “integrity” has a completely different meaning. If I see an “integrity check” script in a cryptography library, I think that its job is to verify that the code is identical to some point of reference.

Also there are other scripts that fall into the same category of sanity checks, and their naming pattern is `check-thing.extension`.

How about `check-files.py`? It isn't very descriptive, but at least it isn't misleading.

gilles-peskine-arm commented on scripts/integrity_check.py:
```diff
@@ -0,0 +1,225 @@
+#
+# Purpose
+#
+# This script requires python 3. This script checks for undesired changes
```
The Python 3 requirement is already indicated by the first line and isn't the most important thing. Please move that sentence to a further paragraph in the description. Also the description should be a file docstring (in `"""`).

gilles-peskine-arm replied:
This script checks the current state, it doesn't check for changes.

mpg commented:
Note: CI failed only on timing test that is know flaky (https://github.com/ARMmbed/mbedtls/issues/1517).

mpg commented:
I'm thinking this is something we probably want to backport, even though it's an enhancement, since it's about testing.

gilles-peskine-arm commented:
https://github.com/ARMmbed/mbedtls/pull/930#pullrequestreview-113247118 (and it's not the first time that's happened) makes me wish for another feature: reject files with lines starting with `<<<<<<`, `=======`, `|||||||` or `>>>>>>>` followed by a space.

sbutcher-arm commented:
> reject files with lines starting with <<<<<<, =======, ||||||| or >>>>>>> followed by a space.

@gilles-peskine-arm - That's a good idea, but let's merge what we have first, and add the feature in later. Raise that as an issue, and we can come back to it.

sbutcher-arm commented:
@dgreen-arm / @mpg - This PR has both 'needs work' and 'needs review', which is a bit confusing. Can you please make clear what work is remaining? And/or whether it's ready for another review?

gilles-peskine-arm commented:
@sbutcher-arm filed as #1589

mpg commented:
@sbutcher-arm I added "needs work" because there is one "change requested" review (yours), and "needs review" because there was no complete second review. If you prefer having only one of these, fine with me, but it had no "needs xxx", which made it look like it was ready for gatekeeper review, which it clearly isn't, so I added something based on what I saw.

@dgreen-arm if you think your latest commit addresses all review feedback so far and you don't plan to do any more work here, please remove the "needs work" label.

dgreen-arm commented:
I've added this to `pre-push.sh`. I noticed that this script is in a different place to all the other checking scripts, so I've moved it to `tests/scripts` for consistency.

I've also rebased this onto the latest development head, so it will need a full re-review.

Patater commented on tests/scripts/check-files.py:
```diff
@@ -0,0 +1,223 @@
+            "minor issues, including incorrect file permissions, "
+            "presence of tabs, non-Unix line endings, trailing whitespace, "
+            "presence of UTF-8 BOM, and TODO comments. "
+            "Note: requires python 3, must be run from Mbed TLS root."
```
Same comment here about running from root

dgreen-arm commented on tests/scripts/check-files.py:
```diff
@@ -0,0 +1,223 @@
+This script checks the current state of the source code for minor issues,
+including incorrect file permissions, presence of tabs, non-Unix line endings,
+trailing whitespace, presence of UTF-8 BOM, and TODO comments.
+Note: requires python 3, must be run from Mbed TLS root.
```
Yes, the script will fail if run elsewhere. This matches the behavior of the other test scripts.
